raise ModelErr when model using Data1DInt have overlapping bins

dtnguyen2 opened the pull request:
# Release Note

The recently released _regrid_ code does not check if the user requested integrated evaluation space have overlap bins. This PR adds a check to make sure the integrated bins do not overlap.  The PR is a fix for issue #569.

DougBurke commented on sherpa/utils/err.py:
```diff
@@ -313,7 +313,7 @@ class ModelErr(SherpaErr):
             'badinstance': '%s instance cannot be created from another model',
             'noncall': 'attempted to create %s from non-callable object of type %s',
-            'needsint': 'An integrated grid is required for model evaluation,\ne.g. [0.1,0.2],[0.2,0.3]',
+            'needsint': 'A non-overlap integrated grid is required for model evaluation,\ne.g. [0.1,0.2],[0.2,0.3]',
```
I think it reads better with 'A non-overlapping '

dtnguyen2 replied:
Done

DougBurke commented on sherpa/models/regrid.py:
```diff
@@ -649,6 +649,11 @@ def _make_and_validate_grid(self, args_array):
+        if self.evaluation_space.is_integrated and requested_eval_space.is_integrated:
+            lo = self.evaluation_space.grid[0]
+            hi = self.evaluation_space.grid[1]
+            if np.any(lo[1:] < hi[:-1]):
```
Do we need to worry about numerical effects here? I am pretty sure that with some of the "does this ARF make sense" checks we could be given grids that slightly overlapped, so there was some effort to support this.

I actually don't think we want to here (i.e. that the existing code is sufficient), but wanted to discuss it. My thought here is that the grid here is almost-certainly user defined (i.e. not from an external definition such as the ARF or RMF) so there's 

 - less likelihood of getting overlapping bins
 - it is much easier for a user to correct the grid here so that it doesn't overlap compared to the ARF/RMF case

anetasie replied:
Agree.

DougBurke commented:
If we are going to make it an error to evaluate the grid with overlapping bins, then shouldn't we make the check in the `regrid` method, when the grids are sent to the model, rather than (or as well as) when the model is actually evaluated?

dtnguyen2 commented:
The fix to issue #569 was in the method called ``` _make_and_validate_grid```, I figured I may as well centralized the sanity check after all the comment does say:
```
        # Ensure the two grids match: integrated or non-integrated.
```
However, while testing the code with the fix in place I noticed the following code:
```
import numpy as np
from sherpa.data import Data1DInt
from sherpa.fit import Fit
from sherpa.models.basic import Box1D

tmp = np.linspace(1, 100, 10)
y = np.ones((9,))
d = Data1DInt('tst', tmp[:-1], tmp[1:], np.ones((9,)))
c = Box1D()
lo = np.linspace(1,100,600)
hi = np.linspace(2,101,600)
c.regrid(lo, hi)
b = Fit(d, c)
r = b.fit()
print(r)
```
does not error out, in fact it returns the following output:
```
datasets       = None
itermethodname = none
methodname     = levmar
statname       = chi2gehrels
succeeded      = True
parnames       = ('box1d.xlow', 'box1d.xhi', 'box1d.ampl')
parvals        = (-0.38540750843724364, 2.237414670577378, 0.808136724942979)
statval        = 1.4826480204147219
istatval       = 1.66797902296655
dstatval       = 0.18533100255182822
numpoints      = 9
dof            = 6
qval           = 0.9606388476822252
rstat          = 0.24710800340245365
message        = successful termination
nfev           = 158
```
I was going to talk with @olaurino about it but I couldn't find him so I figured: may as well submit the PR and raise the issue later etc... but it looks like you beat me to the punch.

anetasie commented:
I'm wondering if this check should apply to the 2D case as well. I'm not sure if there are any restrictions on the 2D model grid generated by the user.

dtnguyen2 commented:
Moved the sanity checks ```_make_and_validate_grid``` to regrid as suggested by @DougBurke, added a low level test to make sure the change does the trick.  

I was hoping to replace the original call to ```_make_and_validate_grid``` to ```EvaluationSpace1D(args)``` or ```EvaluationSpace1D(*args)``` but I couldn't get that to work and not sure why?  Anyone with bright idea

DougBurke commented on sherpa/models/tests/test_regrid_unit.py:
```diff
@@ -448,13 +448,24 @@ def test_regrid1d_non_overlapping_not_allowed():
+
+    tmp = np.linspace(1, 100, 10)
+    y = np.ones((9,))
+    d = Data1DInt('tst', tmp[:-1], tmp[1:], np.ones((9,)))
```
Actually, `d` is unused here, so we don't need `tmp`, `y`, or `d`.

DougBurke commented on sherpa/models/tests/test_regrid_unit.py:
```diff
@@ -437,7 +437,7 @@ def test_regrid1d_error_grid_mismatch_1(setup_1d):
     assert ModelErr.dict['needsint'] in str(excinfo.value)
 
-def test_regrid1d_non_overlapping_not_allowed():
+def test_ui_regrid1d_non_overlapping_not_allowed():
```
Now that the check has been moved to the `regrid` method, there really isn't any difference between `test_ui_regrid1d_non_overlapping_not_allowed` and `test_low_level_regrif1d_non_overlapping_not_allowed`, since both are really just

  - create model instance
  - call regrid on that model instance, check for it raising an error

and the rest of the test (e.g. calling `ui.set_model` or creating a `Data1DInt` instance) isn't adding anything.

dtnguyen2 commented:
Before moving the function ```_make_and_validate_grid``` into ```regrid``` the test ```test_low_level_regrid1d_non_overlapping_not_allowed``` would have failed.  So having the aforementioned test is a requirement for the change to move the sanity check into ```regrid```.  If for some reason someone comes along and and look at the following code in the _sherpa/models/model.py_ file:

```
class RegriddableModel1D(ArithmeticModel):
    def regrid(self, *arrays):
        eval_space = EvaluationSpace1D(*arrays)
        regridder = ModelDomainRegridder1D(eval_space)
        regridder._make_and_validate_grid(arrays)
        return regridder.apply_to(self)

class RegriddableModel2D(ArithmeticModel):
    def regrid(self, *arrays):
        eval_space = EvaluationSpace2D(*arrays)
        regridder = ModelDomainRegridder2D(eval_space)
        return regridder.apply_to(self)
```

and say something like: "hey I want my code to have some symmetry, let's move the call to ```_make_and_validate_grid(arrays)``` back into the ```ModelDomainRegridder1D``` so the two classes look similar then the test will fail.

DougBurke commented:
@dtnguyen2 all I was saying was that I think the two tests are essentially testing the same thing - that they are just

  - create an instance of a model
  - call `regrid` on that instance

I do agree that before the move there was more of a reason for having a low-level and ui-level check.

anetasie commented:
I have issue with getting this to work. Here is a short script:

from sherpa.astro.ui import *
import numpy as np

``` grid_hi = np.linspace(2,102,200)
grid_lo = np.linspace(1,100,200)
grid_hi-grid_lo
grid_hi = np.linspace(1.05,100.05,200)
grid_hi-grid_lo
dataspace1d(1.,100,2,dstype=Data1DInt)
set_model(box1d.b1)
b1.xlow=10
b1.xhi=80
b1.ampl.max=100
b1.ampl=10
fake()
plot_data()
plot_model()

# now I set up model on the higher res grid:
rb1 = b1.regrid(grid_lo,grid_hi)
set_model(rb1)
show_model()
plot_model()  # here I get an error msgs with the Warning,  but the len is the same
len(grid_lo)
len(grid_hi)


In [19]: plot_model()                                                           
/Users/aneta/git/sherpa/sherpa/models/regrid.py:669: UserWarning: evaluation space does not contain the requested space. Sherpa will join the two spaces.
  warnings.warn("evaluation space does not contain the requested space. Sherpa will join the two spaces.")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-19-985f7c7e70f6> in <module>
----> 1 plot_model()

<string> in plot_model(id, **kwargs)

~/git/sherpa/sherpa/astro/ui/utils.py in plot_model(self, id, **kwargs)
  11000             self._plot(id, self._modelhisto, **kwargs)
  11001         else:
> 11002             self._plot(id, self._modelplot, **kwargs)
  11003 
  11004     plot_model.__doc__ = sherpa.ui.utils.Session.plot_model.__doc__

~/git/sherpa/sherpa/ui/utils.py in _plot(self, id, plotobj, *args, **kwargs)
  11744 
  11745         if not sherpa.utils.bool_cast(kwargs.pop('replot', False)):
> 11746             obj = self._prepare_plotobj(id, plotobj, *args)
  11747         try:
  11748             sherpa.plot.begin()

~/git/sherpa/sherpa/astro/ui/utils.py in _prepare_plotobj(self, id, plotobj, resp_id, bkg_id, lo, hi, orders, model)
  10960                 # ids, f = self._get_fit(id)
  10961                 plotobj.prepare(self.get_data(id), self.get_model(id),
> 10962                                 self.get_stat())
  10963 
  10964         return plotobj

~/git/sherpa/sherpa/plot/__init__.py in prepare(self, data, model, stat)
    872     def prepare(self, data, model, stat=None):
    873         (self.x, self.y, self.yerr, self.xerr,
--> 874          self.xlabel, self.ylabel) = data.to_plot(yfunc=model)
    875         self.y = self.y[1]
    876 

~/git/sherpa/sherpa/data.py in to_plot(self, yfunc, staterrfunc)
   1041         # model function, at least in principle.
   1042         return (self.get_x(True, yfunc),
-> 1043                 self.get_y(True, yfunc),
   1044                 self.get_yerr(True, staterrfunc),
   1045                 self.get_xerr(True, yfunc),

~/git/sherpa/sherpa/data.py in get_y(self, filter, yfunc, use_evaluation_space)
    985 
    986         if yfunc is not None:
--> 987             model_evaluation = yfunc(*self.get_evaluation_indep(filter, yfunc, use_evaluation_space))
    988             y = y, model_evaluation
    989 

~/git/sherpa/sherpa/models/model.py in __call__(self, *args, **kwargs)
    262         if (len(args) == 0 and len(kwargs) == 0):
    263             return self
--> 264         return self.calc([p.val for p in self.pars], *args, **kwargs)
    265 
    266     def _get_thawed_pars(self):

~/git/sherpa/sherpa/models/model.py in calc(self, p, *args, **kwargs)
    749 
    750     def calc(self, p, *args, **kwargs):
--> 751         return self.wrapper.calc(p, self.model.calc, *args, **kwargs)
    752 
    753     def get_center(self):

~/git/sherpa/sherpa/models/regrid.py in calc(self, pars, modelfunc, *args, **kwargs)
    624         requested_eval_space = self._make_and_validate_grid(args)
    625 
--> 626         return self._evaluate(requested_eval_space, pars, modelfunc, **kwargs)
    627 
    628     def _make_and_validate_grid(self, args_array):

~/git/sherpa/sherpa/models/regrid.py in _evaluate(self, data_space, pars, modelfunc, **kwargs)
    682                 # This should be the most common case
    683                 y = modelfunc(pars, evaluation_space.grid[0], evaluation_space.grid[1],
--> 684                               **kwargs)
    685                 return rebin(y,
    686                              evaluation_space.grid[0], evaluation_space.grid[1],

~/git/sherpa/sherpa/models/model.py in cache_model(cls, pars, xlo, *args, **kwargs)
     66                 return cache[digest]
     67 
---> 68         vals = func(cls, pars, xlo, *args, **kwargs)
     69 
     70         if use_caching:

~/git/sherpa/sherpa/models/basic.py in calc(self, *args, **kwargs)
    103     def calc(self, *args, **kwargs):
    104         kwargs['integrate'] = bool_cast(self.integrate)
--> 105         return _modelfcts.box1d(*args, **kwargs)
    106 
    107 

TypeError: 1D model evaluation input array sizes do not match, xlo: 249 vs xhi: 250

#  but the arrays have 200 elements - where the 249 and 250 is coming from??

In [20]: len(grid_lo)                                                           
Out[20]: 200

In [21]: len(grid_hi)                                                           
Out[21]: 200

DougBurke commented:
@anetasie - I see the same error message as you get with CIAO 4.11, so it's not an issue with this PR as such. I do see a warning before the error - namely

```
/soft/ciao-4.11/lib/python3.5/site-packages/sherpa-ciao4.11-py3.5-linux-x86_64.egg/sherpa/models/regrid.py:400: UserWarning: evaluation space does not contain the requested space. Sherpa will join the two spaces.
  warnings.warn("evaluation space does not contain the requested space. Sherpa will join the two spaces.")
```

so I think this likely has something to do with the number of bins (the joining of the dataspaces).

I think the issue is that `sherpa.astro.ui.dataspace1d` creates a grid where the last bin is 99-101, and so the regrid model - which only covers 1 to 100.05 - doesn't match, hence the original warning. This then presumably causes a bug to be triggered in a later bit of code.

dtnguyen2 commented:
I think I know where the source of our problem is (ahem it is an existing bug, ie not one added by this PR).  Adding pdb to stop at line 335 of the file _sherpa/models/model.py_

```
(sherpa) [dtn@devel12 sherpa]$ git diff sherpa/models/regrid.py
diff --git a/sherpa/models/regrid.py b/sherpa/models/regrid.py
index 5dbbdec7..57402d31 100644
--- a/sherpa/models/regrid.py
+++ b/sherpa/models/regrid.py
@@ -332,7 +332,8 @@ class EvaluationSpace1D():
             new_xhi = np.unique(np.concatenate((self.x_axis.hi, other.x_axis.hi)))
         else:
             new_xhi = None
-
+        import pdb
+        pdb.set_trace()
         return EvaluationSpace1D(new_xlo, new_xhi)
 
     def __contains__(self, other):
```
So when I run @anetasie (slightly) modified script and stopping to examine the size of new_xlo and new_xhi I get 249 and 250, respectively

```
(sherpa) [dtn@devel12 sherpa]$ python tmp.py
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
/export/sherpa/sherpa/models/regrid.py:670: UserWarning: evaluation space does not contain the requested space. Sherpa will join the two spaces.
  warnings.warn("evaluation space does not contain the requested space. Sherpa will join the two spaces.")
> /export/sherpa/sherpa/models/regrid.py(337)join()
-> return EvaluationSpace1D(new_xlo, new_xhi)
(Pdb) l
332  	            new_xhi = np.unique(np.concatenate((self.x_axis.hi, other.x_axis.hi)))
333  	        else:
334  	            new_xhi = None
335  	        import pdb
336  	        pdb.set_trace()
337  ->	        return EvaluationSpace1D(new_xlo, new_xhi)
338  	
339  	    def __contains__(self, other):
340  	        """
341  	        check if this space properly contains the `other` space, i.e. if the `other` space is contained
342  	        within the boundaries of `self`.
(Pdb) new_xlo.size
249
(Pdb) new_xhi.size
250
```

dtnguyen2 commented:
The method ```join``` in the class ```EvaluationSpace1D``` creates two new arrays called ```new_xlo``` and ```new_xhi``` of size 249 and 250, respectively then instantiate the class ```EvaluationSpace1D``` with two arrays of different sizes but unfortunately the class does not do any sanity checks if the two arrays are of the same size or not.

```
class EvaluationSpace1D():
    """
    Class for 1D Evaluation Spaces. An Evaluation Space is a set of data axes representing
    the data space over which a model can be evaluated.

    A 1D Evaluation Space has only one axis.
    """
    def __init__(self, x=None, xhi=None):
        """
        The input arrays are used to instantiate a single axis.

        Parameters
        ----------
        x : array_like
            The data array, or the low end of the data bins if the dataset is "integrated"
        xhi: array_like
            The high end of the data bins for integrated datasets.
        """
        self.x_axis = Axis(x, xhi)
```

anetasie commented:
ok. The test for non-overlapping bins is fine, but the original issue #569 with plotting the models is still unresolved. Here is the plot_fit() which is correctly showing the fitted model, but the model component plot is just not right (this is the plotting of the model on the finer grid). 

```dataspace1d(1,102.,2,dstype=Data1DInt)
set_model(box1d.b1)
b1.xlow=10
b1.xhi=80
b1.ampl.max=200
b1.ampl=100
fake()
plot_data()
grid_hi2 = np.linspace(1.05,101.05,200)
grid_lo1 = np.linspace(0.55,100.55,200)
grid_hi2 - grid_lo1
rb0=b1.regrid(grid_lo1,grid_hi2)
set_model(rb0)
plot_model()
fit()
plot_fit()
plot_source_component(rb0,overplot=True)

![Figure_1](https://user-images.githubusercontent.com/1213746/68424755-abe48080-0172-11ea-9379-fcedfaf5572b.png)

DougBurke commented:
@anetasie I can not re-create your "wavy" plot - with the following script (I had to guess your dataspace set up call)

```
import numpy as np
from sherpa.ui import *

grid_lo = np.linspace(1,100,200)
grid_hi = np.linspace(1.05,100.05,200)
dataspace1d(1.,100,2,dstype=Data1DInt)

set_model(box1d.b1)
b1.xlow=10
b1.xhi=80
b1.ampl.max=200
b1.ampl=100
fake()

grid_hi2 = np.linspace(1.05,101.05,200)
grid_lo1 = np.linspace(0.55,100.55,200)
rb0=b1.regrid(grid_lo1,grid_hi2)

set_model(rb0)
```

then a

```
plot_fit()
plot_source(overplot=True)
plot_source_component(rb0, overplot=True)
```

gives

![sherpa-regrid](https://user-images.githubusercontent.com/224638/68428153-94f55c80-0179-11ea-8006-aeb7a045bdac.png)

the red line is the `plot_source_component` line, the `orange` line (mostly hidden) is from `plot_fit`, and the green line is from `plot_source(overplot=True)`.

anetasie commented:
Interesting!
I'm getting a good plot first and then after updating the grid I get again the wavy plot.
here is the full screen ouput

```
from sherpa.astro.ui import *
dataspace1d(1,100,2, dstype=Data1dInt)
dataspace1d(1,100,2, dstype=Data1DInt)
set_model(box1d.b1)
b1.xlow=10
b1.xhi=80
b1.ampl.max=100
show_model()
b1.ampl=100
fake()
show_data()
fit()
grid_hi = np.linspace(1.05,101.05,200)
grid_lo = np.linspace(0.55,100.55,200)
rb1 = b1.regrid(grid_lo,grid_hi)
set_model(rb1)
import numpy as np
grid_hi = np.linspace(1.05,101.05,200)
grid_lo = np.linspace(0.55,100.55,200)
rb1 = b1.regrid(grid_lo,grid_hi)
set_model(rb1)
plot_data()
plot_fit()
plot_source()
plot_fit()
plot_source_component(rb1,overplot=True)
fit()
plot_fit()
plot_source_component(rb1,overplot=True)
dataspace1d(1,102.,2,dstype=Data1DInt)
set_model(box1d.b1)
b1.xlow=10
b1.xhi=80
b1.ampl.max=100
show_model()
b1.ampl=100
fake()
b1.ampl.max=200
fake()
set_model(rb1)
plot_data()
fit()
plot_fit()
plot_source_component(rb1,overplot=True)
history

DougBurke commented:
@anetasie - can you open a new issue with your wavy plot example from https://github.com/sherpa/sherpa/pull/685#issuecomment-551274848 so we don't forget it?
